### PR TITLE
Implements context-based matches

### PR DESCRIPTION
**Pull request by tfmar**
This fills in the TODOs left. Fragments generated will now include
context. Also includes a few cleanup/edge cases throughout the
generation code.

### CODE REVIEW

File: test/fragment-generation-utils-test.js [hunk 1]
```diff
@@ -527,4 +527,53 @@ describe('FragmentGenerationUtils', function() {
...
+    target.setStart(node, 5);
+    target.setEnd(node, node.textContent.length - 8);
+    selection.removeAllRanges();
+    selection.addRange(target);
+
+    const result = generationUtils.generateFragment(selection);
+    expect(fragmentUtils.forTesting.normalizeString(result.fragment.prefix))
+        .toEqual('prefix. lorem ipsum dolor');
+    expect(fragmentUtils.forTesting.normalizeString(result.fragment.suffix))
+        .toEqual('recteque qui ei. suffix');
```
Comment (SebLalancette) [inline test/fragment-generation-utils-test.js:577]: Given the length of the selection, does the resulting fragment have textEnd too? Would be nice to assert that here as well, to validate the case where all properties of a fragment would be filled.

File: src/fragment-generation-utils.js [hunk 2]
```diff
@@ -721,7 +743,9 @@ const makeWalkerForNode = (node, endNode) => {
   let blockAncestor = node;
```
Comment (SebLalancette) [inline src/fragment-generation-utils.js:743]: nit: We should probably null-check "node".

**Review by SebLalancette — APPROVED** [review]

File: test/fragment-generation-utils-test.js (hunk 1 above)
Comment (tfmar) [reply test/fragment-generation-utils-test.js:577]: Good call. In testing I found it's a bit fragile to put the exact contents, and that's not really the main thing under test, so I just added a check that it's not falsy.
